File: 10_SimpleTextEditor.py

def simpleTextEditor(operations):
    text = ""
    history = []
    result = []
    for op in operations:
        if op[0] == "1":
            history.append(text)
            text += op[1]
        elif op[0] == "2":
            history.append(text)
            k = int(op[1])
            text = text[:-k]
        elif op[0] == "3":
            result.append(text[int(op[1])-1])
        elif op[0] == "4":
            text = history.pop()
    return result

File: 10_SimpleTextEditor.c

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX 10000

char text[MAX] = "";
char history[100][MAX];
int histTop = -1;

void append(char* s) {
    strcpy(history[++histTop], text);
    strcat(text, s);
}

void delete(int k) {
    strcpy(history[++histTop], text);
    text[strlen(text) - k] = '\0';
}

void print(int k) {
    printf("%c\n", text[k-1]);
}

void undo() {
    strcpy(text, history[histTop--]);
}
